uefi-test-runner/src/proto/network/snp.rs

```diff
@@ -1,40 +1,82 @@
 // SPDX-License-Identifier: MIT OR Apache-2.0
 
-use core::time::Duration;
-
+use core::ops::DerefMut;
+use uefi::proto::device_path::DevicePath;
+use uefi::proto::device_path::text::{AllowShortcuts, DisplayOnly};
 use uefi::proto::network::MacAddress;
-use uefi::proto::network::snp::{InterruptStatus, ReceiveFlags, SimpleNetwork};
+use uefi::proto::network::snp::{InterruptStatus, NetworkState, ReceiveFlags, SimpleNetwork};
 use uefi::{Status, boot};
 
+const ETHERNET_PROTOCOL_IPV4: u16 = 0x0800;
+
+/// Receives the next IPv4 packet and prints corresponding metadata.
+fn receive(simple_network: &mut SimpleNetwork, buffer: &mut [u8]) -> uefi::Result<usize> {
+    let mut recv_src_mac = MacAddress([0; 32]);
+    let mut recv_dst_mac = MacAddress([0; 32]);
+    let mut recv_ethernet_protocol = 0;
+
+    let res = simple_network.receive(
+        buffer,
+        None,
+        Some(&mut recv_src_mac),
+        Some(&mut recv_dst_mac),
+        Some(&mut recv_ethernet_protocol),
+    );
+
+    res.inspect(|_| {
+        debug!("Received:");
+        debug!("  src_mac       =  {:x?}", recv_src_mac);
+        debug!("  dst_mac       =  {:x?}", recv_dst_mac);
+        debug!("  ethernet_proto=0x{:x?}", recv_ethernet_protocol);
```

Comment:
This is asserted below so no need to print.

(In general, I would like to stick to asserting rather than printing as much as possible in tests, because 99% of the time we're not going to look that carefully at the output and won't notice if something doesn't match expectations. For ad-hoc debugging it's easy enough to throw a log line in while working on the code.)

Reply:
The problem is that the `src_mac` is not the one that I configured for the QEMU device. OVMF uses another MAC for reason's I don't understand.

My incentive here is to be as verbose as possible as this test is a minefield IMHO (not because of the way it is written but because of how easy is can fail and how hard it is to modify)